astroid/brain/brain_builtin_inference.py

```diff
@@ -481,6 +481,79 @@ def infer_slice(node, context=None):
     return slice_node
 
 
+def infer_open(node, context=None):  # pylint: disable=unused-argument
+    """Understand `open` (and `file` in Python 2) calls."""
+
+    if six.PY2:
+        code, module_name = _get_py2_open_call_model(node)
+    else:
+        code, module_name = _get_py3_open_call_model(node)
+
+    node = extract_node(code, module_name)
+    return node.instantiate_class()
+
+
+def _get_py2_open_call_model(node):  # pylint: disable=unused-argument
+    # https://docs.python.org/2/library/stdtypes.html#file-objects
+    # https://github.com/python/typeshed/blob/master/stdlib/2/_io.pyi
+    code = dedent('''
+        class file(object): #@
+            closed = True
+            def __enter__(self): return self
+            def __exit__(self, type, value, traceback): return False
+            def __iter__(self): return self
+            def close(self): return None
+            def fileno(self): return 0
+            def flush(self): return None
+            def isatty(self): return True
+            def next(self): return ""
+            def read(self, size=-1): return ""
+            def readline(self, size=-1): return ""
+            def readlines(self, sizehint=-1): return [""]
+            def xreadlines(self, sizehint=-1): yield ""
+            def seek(self, offset, whence=0): return None
+            def tell(self): return 0
+            def truncate(self, size=0): return None
+            def write(self, s): return None
+            def writelines(self, seq): return None
+    ''')
+    module_name = "__builtin__"
+    return code, module_name
+
+
+def _get_py3_open_call_model(node):  # pylint: disable=unused-argument
```

Comment:
I think another good approach, which I considered at some point, is to check the arguments of the ``open`` call. Depending on these, my idea was to return a different object. You can try this idea, in combination with another trick for not regenerating the classes all the time. Instead of building them as you are building them now, you might try to use this:

```
node = extract_node('''
from io import YourIOClass
YourIOClass
''')
inferred = next(n.infer())
return inferred.instantiate_class()
```
What do you think? Can you try this to see if it can work?